src/core/symbology-ng/qgspointdisplacementrenderer.cpp

```diff
@@ -301,6 +308,69 @@ void QgsPointDisplacementRenderer::calculateSymbolAndLabelPositions( QgsSymbolRe
       }
       break;
     }
+    case Grid:
+    {
+      double centerDiagonal = symbolContext.renderContext().convertToPainterUnits( M_SQRT2 * mCenterSymbol->size(),
+                              mCenterSymbol->sizeUnit(), mCenterSymbol->sizeMapUnitScale() );
+      int pointsRemaining = nPosition;
+      gridSize = ceil( sqrt( pointsRemaining ) );
+      if ( pointsRemaining - pow( gridSize - 1, 2 ) < gridSize )
+        gridSize -= 1;
+      double pointRadius = ( centerDiagonal / 2.0 + symbolDiagonal / 2.0 ) + symbolDiagonal;
+
+      int yIndex = 0;
+      while ( pointsRemaining > 0 )
+      {
+        for ( int xIndex = 0; xIndex < gridSize; ++xIndex )
+        {
+          QPointF positionShift( pointRadius * xIndex, pointRadius * yIndex );
+          QPointF labelShift( ( pointRadius + symbolDiagonal / 2 ) * xIndex, ( pointRadius + symbolDiagonal / 2 ) * yIndex );
+          symbolPositions.append( centerPoint + positionShift );
+          labelShifts.append( labelShift );
+          pointsRemaining--;
+        }
+        yIndex++;
+      }
+
+      centralizeGrid( symbolPositions, pointRadius, gridSize );
+      centralizeGrid( labelShifts, pointRadius, gridSize );
+      gridRadius = pointRadius;
+      break;
+    }
+  }
+}
+
+void QgsPointDisplacementRenderer::centralizeGrid( QList<QPointF> &pointSymbolPositions, double radius, int size ) const
+{
+  for ( int i = 0; i < pointSymbolPositions.size(); ++i )
+  {
+    double shiftAmount = -radius * ( size - 1 ) / 2;
+    QPointF centralShift( shiftAmount, shiftAmount );
+    pointSymbolPositions[i] += centralShift;
+  }
+}
+
+void QgsPointDisplacementRenderer::drawGrid( int gridSizeUnits, QgsSymbolRenderContext &context, QList<QPointF> pointSymbolPositions, int nSymbols, double gridRadius )
+{
+  QPainter *p = context.renderContext().painter();
+  if ( nSymbols < 2 || !p ) //draw grid only if multiple features
+  {
+    return;
+  }
+
+  for ( int pointIndex = 0; pointIndex < pointSymbolPositions.size(); ++pointIndex )
```

Comment:
This loop is drawing every inner cell border twice - you can see this by setting a semi-transparent line color and note how the inner borders are darker than the outer ones